Remove old SI DSL dependency and set explicit contentType on output binding

chrisjs opened the pull request.

chrisjs committed:
Remove old SI DSL dependency and set explicit contentType on output binding

artembilan commented on spring-cloud-starter-stream-source-load-generator/src/main/resources/application.yml:
```diff
@@ -0,0 +1,6 @@
+    stream:
+      bindings:
+        output:
+          contentType: application/x-java-serialized-object
```
1. Blank line in the end of each file.
2. We can't enforce `application.yml(properties)` on the library level.

You may consider to do `EnvironmentPostProcessor` (https://github.com/spring-cloud-stream-app-starters/aggregator/blob/master/spring-cloud-starter-stream-processor-aggregator/src/main/java/org/springframework/cloud/stream/app/aggregator/processor/ExcludeStoresAutoConfigurationEnvironmentPostProcessor.java#L36) or somewhere on the programming level. But still: what is the purpose to do this strongly? Can't we somehow make this project JSON-compatible to avoid any reconfiguration and allow end-user to do his/her own decisions about content-type? 

Thanks

chrisjs replied:
* new line added
* are you seeing specific compilation warnings? doing `./mvnw clean install -PgenerateApps` from the top level directory, i don't see any warning indicators
* in regards to application.yml, i was under the impression this is where properties go in boot - can you elaborate on why they can't be enforced? 
* in regards to JSON compatible, i think this goes back to the same discussion we had in another PR with @cppwfs in terms of task-launcher, i had just made the code work with what it was expecting here. if we make one component JSON-compatible it would seem like we would want symmetry across the board which sounds like it would need more discussion (?)

artembilan replied:
Thanks for new lines. That's not critical, indeed, but still it is annoying during review and merge. You can configure your IDE to do that for you automatically on file save.

About warnings: I was just under impression that there are enough deprecation in Boot 2.0, so, some tests might be effected like in your PR for `rabbit`.

Regarding the `application.yml`. Well, we don't develop here the target, finished Spring Boot application. That is end-user responsibility to manage such a configuration externally. We do here just a *starter*. The target self-contained application might be out of our hands. More over we may confuse the scanning system (as well as end-users) with the extra `application.yml` packaged in one or more jars in the classpath.  I really don't have too much arguments now do not do that, but I would stay away of some experiment what will happen in the target application (or SCDF ?) if we package `application.yml` into the jar.

I think what you are going to fix with that `application.yml` is out of `main` code responsibility. That is just a test-case failing. So, and that one we should fix so far. Everything rest regarding JSON-compatibility should be addressed in the separate issue, in the appropriate project. Let's don't mix concerns on the discrete microservice level!

chrisjs replied:
okay - not too sure what "the discreet microservice level" means, but my main goals when starting to make an attempt at fixing these repos was:

1 - make the project compile and tests pass with the types they currently expected where possible (it would feel odd to deserialize JSON or work with it directly in test cases, if a method returns MyType i expect to get MyType)

2 - test produced artifacts making sure they are usable in data flow

i haven't done item 2 for each PR as it would be quite a bit time consuming and I'm relying on the individual project tests/patterns, but  item 1 has been part of all.

so with that said, and correct me if I am wrong -- i think what you are saying is "don't force a contentType in starter application.yml files but forcing conversion to strict type in test properties is fine as thats what the test is expecting (or wants to work with)" -- would that be correct? if so then i think i see better where you are going with this -- the JSON default is sent over the wire and left for the client to deal with be it by deserializing how they wish via code or by setting a contentType property to have conversion done for them. 

AFAICT from what I have seen so far, the downside of letting JSON fly as the default and not setting a contentType to return what the app used to (by now forcing in application.yml) is things break on the data flow side as types are generally expected over JSON currently. but i suppose the "fix" for that is to set a property to define the expected content type is by the client (i.e.: in DF shell) and that is the separation between a "stream app" and a "data flow app"?

chrisjs replied:
okay - not too sure what "the discreet microservice level" means, but my main goals when starting to make an attempt at fixing these repos was:

1 - make the project compile and tests pass with the types they currently expected where possible (it would feel odd to deserialize JSON or work with it directly in test cases, if a method returns MyType i expect to get MyType)

2 - test produced artifacts making sure they are usable in data flow

i haven't done #2 for each PR as it would be quite a bit time consuming and I'm relying on the individual project tests/patterns, but #1 has been part of all.

so with that said, and correct me if I am wrong -- i think what you are saying is "don't force a contentType in starter application.yml files but forcing conversion to strict type in test properties is fine as thats what the test is expecting (or wants to work with)" -- would that be correct? if so then i think i see better where you are going with this -- the JSON default is sent over the wire and left for the client to deal with be it by deserializing how they wish via code or by setting a contentType property to have conversion done for them. 

AFAICT from what I have seen so far, the downside of letting JSON fly as the default and not setting a contentType to return what the app used to (forcing in application.yml) is things break on the data flow side as types are generally expected over JSON currently. but i suppose the "fix" for that is to set a property to define the expected content type is by the client (i.e.: in DF shell) and that is the separation between a "stream app" and a "data flow app"?

chrisjs replied:
ok so i updated the PR to remove the `application.yml` and move the content type setting into the test case otherwise the project wouldn't build. i used `application/octet-stream` since thats what the README indicates it should be and the test passes.

when used in data flow along with `throughput`, for example:

`stream create --name testLoad --definition "load-generator | throughput" --deploy`

the `throughput` sink receives messages properly. i am assuming this works out of the box because https://github.com/spring-cloud-stream-app-starters/throughput/blob/master/spring-cloud-starter-stream-sink-throughput/src/main/java/org/springframework/cloud/stream/app/throughput/sink/ThroughputSinkConfiguration.java#L84 -- the method signature is `Message<?>` and marked with `@ServiceActivator`. it seems according to https://docs.spring.io/spring-integration/api/org/springframework/integration/annotation/ServiceActivator.html I'm guessing thats why a JSON input is being converted automatically into a `Message` in this case

artembilan reviewed (changes requested):
Re-check also, please, for compilation warnings

artembilan commented:
Thank you for the update, @chrisjs !

You know my main concern here just to produce a self-contained Microservice, not tied to any target platform. I would even try to stay away from mentioning SCDF here at all and don't try to add struts in our code to make that happy if any. 

On the other hand if we have some domain object not compatible with possible content types (end-user can decide what he/she prefers to use), we have to fix that one. And still don't bring any workaround to main code. In the end we just a Microservice with its one contract! According to the latest wire protocol we have a JSON by default, so the test case must be updated accordingly. 

Just IMHO. Would be great to hear opinion of other guys as well.

/CC @sabbyanandan , @sobychacko

chrisjs commented:
so just the test case was updated to reflect the content type this app expects according to https://github.com/spring-cloud-stream-app-starters/load-generator/blob/master/spring-cloud-starter-stream-source-load-generator/README.adoc -- is that sufficient here or you looking for the tests to deserialize the default payload (JSON) and work against that?

sobychacko commented:
@artembilan @chrisjs I see that the latest update in the PR only updates the `contentType` in the test, and keeping the default `contentType` of `application/json` for any users of the app. While it might be fine, I am not sure how it works with downstream consumers of the app - i.e. if they are expecting the app to work  in a seamless manner as in 1.3, it won't be the case - they have to set the content type explicitly now.

artembilan commented:
I'm fine with either way, just not the `main` code base workaround.
Right, in other project I just added Jackson `ObjectMapper` to deserialize the content on the wire.

Good catch regarding the README though! 

Well, now I'm confused. According the latest SCSt we always have JSON as a content type by default. So, to follow that README we always have to enforce end-users to set the content type to the `application/octet-stream`. Although we could change that README as well and say that we produce JSON by default now as well. That's how SCSt 2.0 works, nothing more.

artembilan commented:
Looking to this one more time, I'm going to stay away of merging. I don't feel like I agree with one or another way to make it working:

1. We explicitly say that we we support only `application/octet-stream`, but at the same time the SCSt 2.0 infrastructure converts the `payload` to the `application/json`. Inconsistency.
2. To make the test working and follow the declared `application/octet-stream` we specify the `contentType`. Seems for me that's not what end-users would expect from us. Although we have bumped not one time already and taught people to live with default JSON or configure expected `contentType`. Inconvenience.
3. We could change the declared `contentType` to JSON as it is the default for the SCSt 2.0. But is that really a JSON what we produce here? Confusing.

As a compromise I would suggest to do an explicit `contentType = application/octet-stream` in the `EnvironmentPostProcessor`, but only if it isn't there yet. This way we would follow declared `application/octet-stream` in the README, but staid away of the hidden `application.yml`.

I need to hear an opinion of some one else, or just take this PR from me and merge it by your vision...

artembilan commented:
@chrisjs ,

OK. Let me come back to you with the decision we did with the team!
@sobychacko , please, correct me if I'm wrong somewhere.

We definitely need to go the way for the content we expect here.
So, since we declare `application/octet-stream` as a recommended `contentType`, let's set it by default.
The `EnvironmentPostProcessor` should be used for this purpose. And we don't override any explicit end-user `contentType` settings.

I guess this is the rule for all our app starters where we don't deal with any arbitrary `contentType`.

We can't provide the `application.yml` because there is no guarantee which one is going to be load last to be able to provide the final value for the property. I mean that we still have to provide the freedom for end-users for default property overriding.

Hope that's clear.

chrisjs commented:
ill be going the EnvironmentPostProcessor path you mentioned, using the content-type specified by the project README if the property has not already been set.

chrisjs committed:
Use EnvironmentPostProcessor rather than using test properties or application.yml

chrisjs commented:
removed setting of test property and added a EnvironmentPostProcessor. I'm assuming there will be other app projects that will be needing to do the same, so maybe this will end up more generic and shared. ill revisit when i hit another similar one scenario again.

artembilan reviewed (approved):
Sounds good, @chrisjs !

Merging...
